plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeInsightsSettingsFacade.kt

```diff
@@ -0,0 +1,98 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.codewhisperer.util
+
+import com.intellij.codeInsight.CodeInsightSettings
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.util.Disposer
+import com.intellij.openapi.util.SimpleModificationTracker
+import org.jetbrains.annotations.VisibleForTesting
+import software.aws.toolkits.core.utils.error
+import software.aws.toolkits.core.utils.getLogger
+import kotlin.reflect.KMutableProperty
+
+class CodeInsightsSettingsFacade : SimpleModificationTracker(), Disposable {
+    private inner class ChangeAndRevert<T : Any>(
+        val p: KMutableProperty<T>,
+        val value: T,
+        parentDisposable: Disposable
+    ) {
+        val origin: T = p.getter.call()
+        var isComplete: Boolean = false
+            private set
+
+        init {
+            Disposer.register(parentDisposable) {
+                revert()
+            }
+        }
+
+        fun commit(): ChangeAndRevert<T> {
+            p.setter.call(value)
+            return this
+        }
+
+        fun revert() {
+            if (isComplete) {
+                return
+            }
+
+            p.setter.call(origin)
+            isComplete = true
+        }
+    }
+
+    private val settings by lazy {
+        CodeInsightSettings.getInstance()
+    }
+
+    private var pendingReverts = listOf<ChangeAndRevert<*>>()
+    val pendingRevertCounts: Int
+        get() = pendingReverts.size
+
+    @VisibleForTesting
+    internal fun revertAll() {
+        if (pendingReverts.count { !it.isComplete } == 0) {
+            return
+        }
+
+        pendingReverts.forEach {
+            it.revert()
+        }
+
+        pendingReverts = pendingReverts.filter {
+            !it.isComplete
+        }.toMutableList()
+    }
+
+    fun disableCodeInsightUntil(parentDisposable: Disposable) {
+        revertAll()
+        val toReverts = mutableListOf<ChangeAndRevert<*>>()
+
+        ChangeAndRevert(settings::TAB_EXITS_BRACKETS_AND_QUOTES, false, parentDisposable).commit().also {
+            toReverts.add(it)
+        }
+
+        ChangeAndRevert(settings::AUTO_POPUP_COMPLETION_LOOKUP, false, parentDisposable).commit().also {
+            toReverts.add(it)
+        }
```

Comment:
is this section short? is it possible for a user to change these settings while code insight is temporarily disabled?

Reply:
yea it's short and plus codewhisperer popups will be closed if they're losing focus, so in most cases, popup will be closed and reverted when users opening up the IDE settings panel to configure their CodeInsightSettings. But worth checking out if we have a way to handle that corner case. Will take it as a followup after clearing out other high priority issues.



https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/96e03915-79b7-426d-9837-d4c03b7c8f15